filesystem/filesystem.go

```diff
@@ -496,10 +502,75 @@ func (m *Mount) addMetadata(path string, md metadata.Metadata, owner *user.User)
 	return m.writeDataAtomic(path, data, owner)
 }
 
+func readFull(fd int, data []byte) (int, error) {
+	i := 0
+	for i < len(data) {
+		count, err := unix.Read(fd, data[i:])
+		if err != nil {
+			return 0, err
+		}
+		if count == 0 {
+			break
+		}
+		i += count
+	}
+	return i, nil
+}
+
+// readMetadataFileSafe gets the contents of a metadata file extra-carefully,
+// considering that it could be a malicious file created to cause a
+// denial-of-service.  Specifically, the following checks are done:
+//
+// - It must be a regular file, not another type of file like a symlink or FIFO.
+// - It must have a reasonable size (<= maxMetadataFileSize).
+//
+// Take care to avoid TOCTOU (time-of-check-time-of-use) bugs when doing these
+// tests.  Notably, we must open the file before checking the file type, as the
+// file type could change between any previous checks and the open.  When doing
+// this, O_NOFOLLOW is needed to avoid following a symlink, and O_NONBLOCK is
+// needed to avoid blocking if the file is a FIFO.  Note that this implies we
+// have to use the low-level unix.Open instead of os.OpenFile.
+//
+// This function returns the data read.
+func readMetadataFileSafe(path string) ([]byte, error) {
+	fd, err := unix.Open(path, unix.O_RDONLY|unix.O_NOFOLLOW|unix.O_NONBLOCK, 0)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to open %q", path)
+	}
+	defer unix.Close(fd)
+
+	var stat unix.Stat_t
+	if err = unix.Fstat(fd, &stat); err != nil {
+		return nil, errors.Wrapf(err, "failed to stat %q", path)
```

Comment:
Here and elsewhere, if we have to use raw errors, we should return a [`os.PathError`](https://pkg.go.dev/os#PathError)

Reply:
Done in the one place that still uses `unix` directly (`unix.FnctlInt`)